streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplier.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.TimestampedKeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+
+public class ForeignJoinSubscriptionProcessorSupplier<K, KO, VO> implements ProcessorSupplier<KO, Change<VO>> {
+    private static final Logger LOG = LoggerFactory.getLogger(ForeignJoinSubscriptionProcessorSupplier.class);
+    private final StoreBuilder<TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>>> storeBuilder;
+    private final CombinedKeySchema<KO, K> keySchema;
+
+    public ForeignJoinSubscriptionProcessorSupplier(
+        final StoreBuilder<TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>>> storeBuilder,
+        final CombinedKeySchema<KO, K> keySchema) {
+
+        this.storeBuilder = storeBuilder;
+        this.keySchema = keySchema;
+    }
+
+    @Override
+    public Processor<KO, Change<VO>> get() {
+        return new KTableKTableJoinProcessor();
+    }
+
+
+    private final class KTableKTableJoinProcessor extends AbstractProcessor<KO, Change<VO>> {
+        private Sensor skippedRecordsSensor;
+        private TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>> store;
+
+        @Override
+        public void init(final ProcessorContext context) {
+            super.init(context);
+            final InternalProcessorContext internalProcessorContext = (InternalProcessorContext) context;
+            skippedRecordsSensor = ThreadMetrics.skipRecordSensor(internalProcessorContext.metrics());
+            store = internalProcessorContext.getStateStore(storeBuilder);
+        }
+
+        /**
+         * @throws StreamsException if key is null
+         */
+        @Override
+        public void process(final KO key, final Change<VO> value) {
+            // if the key is null, we do not need proceed aggregating
+            // the record with the table
+            if (key == null) {
+                LOG.warn(
+                    "Skipping record due to null key. value=[{}] topic=[{}] partition=[{}] offset=[{}]",
+                    value, context().topic(), context().partition(), context().offset()
+                );
+                skippedRecordsSensor.record();
+                return;
+            }
+
+            final Bytes prefixBytes = keySchema.prefixBytes(key);
+
+            //Perform the prefixScan and propagate the results
+            try (final KeyValueIterator<Bytes, ValueAndTimestamp<SubscriptionWrapper<K>>> prefixScanResults =
+                     store.range(prefixBytes, Bytes.increment(prefixBytes))) {
+
+                while (prefixScanResults.hasNext()) {
+                    final KeyValue<Bytes, ValueAndTimestamp<SubscriptionWrapper<K>>> next = prefixScanResults.next();
+                    // have to check the prefix because the range end is inclusive :(
```

Comment:
This is unfortunate, but I think it's better to avoid changing the Store interfaces at this time.

Reply:
Yep, but the comment makes it clear.